### DISCUSSION TITLE
Investigate IME-plugin integration on Linux

### DISCUSSION THREAD

**Discussion by federico-terzi** [espanso/espanso#1247] [Feature requests and Ideas]
As proposed by @eternaleye, integrating Espanso as an IME plugin might offer many advantages on Linux.

I'm posting the great comments by @eternaleye here:

Any thoughts on integrating on Linux via providing an IME plugin, either via existing protocols (display-protocol-native like XIM/wl-text-input, or existing IME services like ibus/fcitx5) or via toolkit plugins (GTK and Qt both have IME plugin interfaces that give you a lot of control)?

I ask because sniffing all input isn't especially portable (on Wayland, it's explicitly blocked as a security issue), but people do already use IME engines for similar use cases (in my case, I have an emoji database and a LaTeX sequence database loaded into fcitx5's table plugin), and their main weakness is exactly where Espanso is strong: they lack a large, easy-to-use corpus of drop-in databases.

> Thank you for the feedback! This sounds interesting, could please you point me to some open-source projects in that area? I'd love to investigate how those are working under the hoods

Sure! So, all of the things I named are open-source:

- [XIM, the X Input Method](https://www.x.org/releases/X11R7.6/doc/libX11/specs/XIM/xim.html) is part of the X11 protocol, to see how to implement an input method that uses it, you can see how [fcitx5](https://github.com/fcitx/fcitx5/tree/master/src/frontend/xim) or [ibus](https://github.com/ibus/ibus/tree/dd4cc5b028c35f9bb8fa9d3bdc8f26bcdfc43d40/client/x11) do so. If you support XIM, users who are running under X11 will near-universally be able to use your software, but it does nothing on Wayland. It also has somewhat limited functionality, and potentially poor-looking UI.
- [wl-text-input](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/tree/main/unstable/text-input) has the same relationship to the Wayland display protocol as XIM does to X11, and both [fcitx5](https://github.com/fcitx/fcitx5/tree/master/src/frontend/waylandim) and [ibus](https://github.com/ibus/ibus/tree/dd4cc5b028c35f9bb8fa9d3bdc8f26bcdfc43d40/client/wayland) both implement it, if you want to see how. Converse to XIM, this will give you near-universal support on Wayland, but nothing on X11. It's also quite new.

As for toolkits, both GTK and Qt support input methods that work via the above protocols, _but_ you can get tighter integration (and more functionality) by directly implementing the toolkit-specific input method plugin interface. However, be aware - major toolkit versions need separate plugins. Qt 5 and GTK 3 are currently the most widely deployed, but both Qt 6 and GTK 4 are going to be much more relevant before long.
- In Qt, the docs on QPlatformInputContext are relatively bare, but KDAB did a [reasonably deep dive](https://www.kdab.com/qt-input-method-depth/) on the subject as well. In addition, [fcitx5](https://github.com/fcitx/fcitx5-qt/tree/master/qt5/platforminputcontext) and [ibus](https://github.com/ibus/ibus-qt/tree/master/qtim) have implemented Qt input method plugins. Supporting this will get you solid integration in any Qt-based application, including all KDE programs.
- In GTK, the [upstream docs](https://docs.gtk.org/gtk3/class.IMContext.html) are a bit more helpful, and both [fcitx5](https://github.com/fcitx/fcitx5-gtk/tree/master/gtk3) and [ibus](https://github.com/ibus/ibus/tree/dd4cc5b028c35f9bb8fa9d3bdc8f26bcdfc43d40/client/gtk3) have implemented such plugins. This will get you solid integration in Gtk-based applications, such as Inkscape, Firefox, and Gnome-based applications.

Meanwhile, [fcitx5](https://github.com/fcitx/fcitx5) and [ibus](https://github.com/ibus/ibus) are basically frameworks to make it both easier to write input methods _and_ to make any that you write more portable, by acting as a common interface (and single daemon) that integrates with all the platform-specific interfaces (such as XIM or wl-text-input) and all the toolkits (such as GTK or Qt). As you likely noticed in the above entries, implementing a plugin for either of these gets you all the others basically for free. In addition, you can ride on their existing documentation for getting the system integrations all set up, and they have GUI configurators for enabling their modules (unlike the other options). Another upside is that if people already use one of these for other reasons (such as CJK input), they won't be forced to choose between that and espanso - these support multiple plugins being active. I personally use fcitx5, but I used to use ibus and it has overall higher usage in the community as far as I know.
- For fcitx5, you may want to look at [fcitx5-anthy](https://github.com/fcitx/fcitx5-anthy), a Japanese module for fcitx5 using the external Anthy library (as it shows integration of a third-party library with fcitx5 as an input method provider)
- For ibus, it similarly has an [ibus-anthy](https://github.com/ibus/ibus-anthy) module, which should enable apples-to-apples comparison.

I hope this helps!

**Comment by federico-terzi** [comment]
Possible starting point: https://github.com/ibus/ibus-tmpl
Also, This one seems to be a good starting point: https://github.com/salty-horse/ibus-uniemoji/blob/master/ibus.py

Also this project is great, though a bit complex to extract only the relevant bits: https://github.com/mike-fabian/ibus-typing-booster

**Comment by eternaleye** [comment]
If you're interested in looking at the closest fcitx5 has to espanso currently, you might want to check out the [QuickPhrase](https://github.com/fcitx/fcitx5/tree/master/src/modules/quickphrase) module that ships with the core.

It's what I use myself; it comes with an [emoji database](https://github.com/fcitx/fcitx5/blob/master/src/modules/quickphrase/quickphrase.d/emoji-eac.mb) and a [LaTeX escape database](https://github.com/fcitx/fcitx5/blob/master/src/modules/quickphrase/quickphrase.d/latex.mb), and there _is_ a GUI in the fcitx5 UI for adding or removing entries to custom databases, which live in `${XDG_DATA_HOME}/fcitx5/data/quickphrase.d/`. Nothing like espanso's package manager, though.

I personally use their emoji DB, a customized LaTeX DB (to have more codes, mostly), and an XCompose sequence database I converted using a script I wrote. I have my prefix key to activate it set to CapsLock, so I can do `<CapsLock>\forall` and get ∀.

EDIT: Some screenshots:

![image](https://user-images.githubusercontent.com/99404/147377874-0f671526-caf0-40ca-97c4-51f62ed75068.png)

![image](https://user-images.githubusercontent.com/99404/147377880-4f523740-1dd2-45dc-9369-13df1df7dd7c.png)

**Comment by federico-terzi** [comment]
@eternaleye Thanks! That looks promising

Yesterday I've experimented a bit with IBus to understand the main dynamics and, _in theory_, an integration should be possible. I suspect that with fcitx5 the situation would be similar. 

The biggest obstacle to overcome is figuring out how to (somewhat cleanly) integrate with the new Espanso architecture, which is asynchronous, compared to a regular IME program which (as far as my limited knowledge goes) is synchronous.

I'm particularly excited about the possible improvements the IME integration could bring on Wayland

I can't promise anything, but I'll keep investigating :) Thanks again for the help!

**Comment by eternaleye** [comment]
No problem, glad to help!
